src/jit/lowerxarch.cpp

```diff
@@ -2299,6 +2299,18 @@ void Lowering::TreeNodeInfoInitLogicalOp(GenTree* tree)
         // as reg optional.
         SetRegOptionalForBinOp(tree);
     }
+
+    if (varTypeIsByte(op1))
+    {
+        regMaskTP regMask;
+        bool      containedNode = (op2->gtLsraInfo.srcCount == 0) && (op2->gtLsraInfo.dstCount == 0);
+        if (!containedNode)
+        {
+            regMask = op2->gtLsraInfo.getSrcCandidates(l);
+            assert(regMask != RBM_NONE);
+            op2->gtLsraInfo.setSrcCandidates(l, regMask & ~RBM_NON_BYTE_REGS);
+        }
```

Comment:
I believe (and I don't feel that I grok this 100%) it is the case that these are duals - i.e. you never have both. The idea is that you count on either having a full integer-width entity that you can load from and store into (normalize on store), OR you have to load from and store into a byte and zero- or sign-extend as needed on load (because it's a field of a local struct or an arg). In the former case you need to make sure that you only store a byte's worth, even if you have somehow accumulated garbage in the upper bits. In the latter case, it doesn't matter because you're only storing a byte. I hope somebody will clue me in if I don't have this right!